pkg/apis/serving/v1alpha2/kfservice_types.go

```diff
@@ -21,36 +21,73 @@ import (
-	// Canary defines an alternate configuration to route a percentage of traffic.
-	Canary               *ModelSpec `json:"canary,omitempty"`
-	CanaryTrafficPercent int        `json:"canaryTrafficPercent,omitempty"`
+	// Default defines default KFService endpoints
+	// +required
+	Default EndpointSpec `json:"default"`
+	// Canary defines an alternate endpoints to route a percentage of traffic.
+	// +optional
+	Canary *EndpointSpec `json:"canary,omitempty"`
+	// CanaryTrafficPercent defines the percentage of traffic going to canary KFService endpoints
+	// +optional
+	CanaryTrafficPercent int `json:"canaryTrafficPercent,omitempty"`
 }
 
-// ModelSpec defines the configuration to route traffic to a predictor.
-type ModelSpec struct {
+type EndpointSpec struct {
+	// Predictor defines the model serving spec
+	// +required
+	Predictor PredictorSpec `json:"predictor"`
+
+	// Explainer defines the model explanation service spec
+	// explainer service calls to transformer or predictor service
+	// +optional
+	Explainer *ExplainerSpec `json:"explainer,omitempty"`
+
+	// Transformer defines the transformer service spec for pre/post processing
+	// transformer service calls to predictor service
+	// +optional
+	Transformer *TransformerSpec `json:"transformer,omitempty"`
+}
+
+// DeploymentSpec defines the configuration for a given KFService service component
+type DeploymentSpec struct {
+	// ServiceAccountName is the name of the ServiceAccount to use to run the service
+	// +optional
 	ServiceAccountName string `json:"serviceAccountName,omitempty"`
 	// Minimum number of replicas, pods won't scale down to 0 in case of no traffic
+	// +optional
 	MinReplicas int `json:"minReplicas,omitempty"`
 	// This is the up bound for autoscaler to scale to
+	// +optional
 	MaxReplicas int `json:"maxReplicas,omitempty"`
+}
+
+// PredictorSpec defines the configuration to route traffic to a predictor.
+type PredictorSpec struct {
 	// The following fields follow a "1-of" semantic. Users must specify exactly one spec.
 	Custom     *CustomSpec     `json:"custom,omitempty"`
 	Tensorflow *TensorflowSpec `json:"tensorflow,omitempty"`
 	TensorRT   *TensorRTSpec   `json:"tensorrt,omitempty"`
 	XGBoost    *XGBoostSpec    `json:"xgboost,omitempty"`
 	SKLearn    *SKLearnSpec    `json:"sklearn,omitempty"`
 	PyTorch    *PyTorchSpec    `json:"pytorch,omitempty"`
-	// Optional Explain specification to add a model explainer next to the chosen predictor.
-	// In future v1alpha2 the above model predictors would be moved down a level.
-	Explain *ExplainSpec `json:"explain,omitempty"`
+
+	DeploymentSpec `json:",inline"`
 }
 
-// ExplainSpec defines the arguments for a model explanation server
-type ExplainSpec struct {
+// ExplainerSpec defines the arguments for a model explanation server
+type ExplainerSpec struct {
 	// The following fields follow a "1-of" semantic. Users must specify exactly one spec.
 	Alibi  *AlibiExplainSpec `json:"alibi,omitempty"`
```

Comment:
Should `AlibiExplainSpec` be changed to `AlibiExplainerSpec` to fit the agreed noun convention?